htlcswitch/switch.go

```diff
@@ -345,12 +335,33 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
 func (s *Switch) GetPaymentResult(paymentID uint64, paymentHash lntypes.Hash,
 	deobfuscator ErrorDecrypter) (<-chan *PaymentResult, error) {
 
-	s.pendingMutex.Lock()
-	payment, ok := s.pendingPayments[paymentID]
-	s.pendingMutex.Unlock()
+	var (
+		nChan  <-chan *networkResult
+		err    error
+		outKey = CircuitKey{
+			ChanID: sourceHop,
+			HtlcID: paymentID,
+		}
+	)
 
-	if !ok {
-		return nil, ErrPaymentIDNotFound
+	// If the payment is not found in the circuit map, check whether a
+	// result is already available.
+	// Assumption: no one will add this payment ID other than the caller.
+	if s.circuits.LookupCircuit(outKey) == nil {
+		res, err := s.networkResults.getResult(paymentID)
+		if err != nil {
+			return nil, err
+		}
+		c := make(chan *networkResult, 1)
+		c <- res
+		nChan = c
+	} else {
+		// The payment was committed to the circuits, subscribe for a
+		// result.
+		nChan, err = s.networkResults.subscribeResult(paymentID)
```

Comment:
With this new approach, it seems that a caller that passes in an unknown `paymentID` will block indefinitely, rather than receiving an immediate error. I'm guessing it was done this way in order to avoid storing mostly duplicate data here and also in the router?

Reply:
If the payment ID is unknown, `s.circuits.LookupCircuit(outKey) == nil` above will be true, and we attempt to fetch the result directly. `ErrUnknownPaymetnID` will be returned.